stdlib/src/utils/string_slice.mojo

```diff
@@ -26,6 +26,155 @@ alias StaticString = StringSlice[ImmutableStaticLifetime]
+            for i in range(1, width):
+                if byte_types[i] != 0:
+                    idx += i
+                    iter_len -= i
+                    break
+            continue
+
+        # byte_type has to match against the amount of continuation bytes
+        alias vec_t = SIMD[DType.uint8, 4]
+        alias n4_byte_types = vec_t(4, 1, 1, 1)
+        alias n3_byte_types = vec_t(3, 1, 1, 0)
+        alias n3_mask = vec_t(0b111, 0b111, 0b111, 0)
+        alias n2_byte_types = vec_t(2, 1, 0, 0)
+        alias n2_mask = vec_t(0b111, 0b111, 0, 0)
+        var byte_types_4 = byte_types.slice[4]()
+        var valid_n4 = (byte_types_4 == n4_byte_types).reduce_and()
+        var valid_n3 = ((byte_types_4 & n3_mask) == n3_byte_types).reduce_and()
+        var valid_n2 = ((byte_types_4 & n2_mask) == n2_byte_types).reduce_and()
+        if not (valid_n4 or valid_n3 or valid_n2):
+            return -1
+
+        # special unicode ranges
+        var b0 = d[0]
+        var b1 = d[1]
+        if first_byte_type == 2 and b0 < UInt8(0b1100_0010):
+            return -1
+        elif b0 == 0xE0 and not (UInt8(0xA0) <= b1 <= UInt8(0xBF)):
+            return -1
+        elif b0 == 0xED and not (UInt8(0x80) <= b1 <= UInt8(0x9F)):
+            return -1
+        elif b0 == 0xF0 and not (UInt8(0x90) <= b1 <= UInt8(0xBF)):
+            return -1
+        elif b0 == 0xF4 and not (UInt8(0x80) <= b1 <= UInt8(0x8F)):
+            return -1
+
+        # amount of bytes evaluated
+        idx += int(first_byte_type)
+        iter_len -= int(first_byte_type)
+
+        @parameter
+        if remainder:
+            break
+    return iter_len
+
+
+fn _is_valid_utf8(data: UnsafePointer[UInt8], length: Int) -> Bool:
+    """Verify that the bytes are valid UTF-8.
+
+    Args:
+        data: The pointer to the data.
+        length: The length of the items pointed to.
+
+    Returns:
+        Whether the data is valid UTF-8.
+
+    #### UTF-8 coding format
+    [Table 3-7 page 94](http://www.unicode.org/versions/Unicode6.0.0/ch03.pdf).
+    Well-Formed UTF-8 Byte Sequences
+
+    Code Points        | First Byte | Second Byte | Third Byte | Fourth Byte |
+    :----------        | :--------- | :---------- | :--------- | :---------- |
+    U+0000..U+007F     | 00..7F     |             |            |             |
+    U+0080..U+07FF     | C2..DF     | 80..BF      |            |             |
+    U+0800..U+0FFF     | E0         | ***A0***..BF| 80..BF     |             |
+    U+1000..U+CFFF     | E1..EC     | 80..BF      | 80..BF     |             |
+    U+D000..U+D7FF     | ED         | 80..***9F***| 80..BF     |             |
+    U+E000..U+FFFF     | EE..EF     | 80..BF      | 80..BF     |             |
+    U+10000..U+3FFFF   | F0         | ***90***..BF| 80..BF     | 80..BF      |
+    U+40000..U+FFFFF   | F1..F3     | 80..BF      | 80..BF     | 80..BF      |
+    U+100000..U+10FFFF | F4         | 80..***8F***| 80..BF     | 80..BF      |
+    .
```

Comment:
nit: Please move this before the `Args:`.

Reply:
I see, I never used a footnote in a docstring, I will see if it breaks anything on our end.